### PR TITLE
gh-89928: Fix integer conversion of device numbers.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Fix os.major(), os.minor() and os.makedev().
Support device numbers larger than 2**63-1.
Support non-existent device number (NODEV).

https://bugs.python.org/issue45767

* Issue: gh-89928

### CODE REVIEW

**Review by iritkatriel — DISMISSED** [review]
This has merge conflicts now.

**Comment by serhiy-storchaka** [comment]
!buildbot freebsd

**Comment by serhiy-storchaka** [comment]
I have made the requested changes; please review again.

**Comment by serhiy-storchaka** [comment]
!buildbot bsd

File: Modules/posixmodule.c [hunk 1]
```diff
@@ -12118,53 +12149,77 @@ os_mknod_impl(PyObject *module, path_t *path, int mode, dev_t device,
...
 
-static unsigned int
+static PyObject *
 os_major_impl(PyObject *module, dev_t device)
-/*[clinic end generated code: output=5b3b2589bafb498e input=1e16a4d30c4d4462]*/
+/*[clinic end generated code: output=4071ffee17647891 input=b1a0a14ec9448229]*/
 {
-    return major(device);
+    unsigned int result = major(device);
+#ifdef NODEV
```
Comment (vstinner) [inline Modules/posixmodule.c]: Can you add a helper function to convert an "unsigned int" device number to a Python object? os.minor() uses the same code.

File: Misc/NEWS.d/next/Library/2022-03-10-16-47-57.bpo-45767.ywmyo1.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Fix integer conversion in :func:`os.major`, :func:`os.minor`, and
+:func:`os.makedev`. Support device numbers larger than ``2**63-1``. Support
+non-existent device number.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2022-03-10-16-47-57.bpo-45767.ywmyo1.rst]:
```suggestion
non-existent device number (`NODEV`).
```

**Comment by vstinner** [comment]
I closed https://github.com/python/cpython/pull/113080 as a duplicate of this PR.

File: Lib/test/test_posix.py [hunk 3]
```diff
@@ -704,21 +704,32 @@ def test_makedev(self):
...
         self.assertRaises(TypeError, posix.makedev, float(major), minor)
         self.assertRaises(TypeError, posix.makedev, major, float(minor))
         self.assertRaises(TypeError, posix.makedev, major)
         self.assertRaises(TypeError, posix.makedev)
+        for x in -2, 2**32, 2**64, -2**63-1:
+            self.assertRaises((ValueError, OverflowError), posix.makedev, x, minor)
+            self.assertRaises((ValueError, OverflowError), posix.makedev, major, x)
+
+        if sys.platform == 'linux':
+            NODEV = -1
```
Comment (serhiy-storchaka) [inline Lib/test/test_posix.py:729]: I am going to expose `posix.NODEV`, but this is a different issue.

**Review by vstinner — APPROVED** [review]
LGTM. Thanks for the update.

**Comment by vstinner** [comment]
@serhiy-storchaka: Do you want to merge this PR?

**Comment by vstinner** [comment]
Thanks for the fix @serhiy-storchaka!

File: Lib/test/test_posix.py (hunk 3 above)
Comment (kraj) [reply Lib/test/test_posix.py:729]: FWIW, musl is another C library implementation on Linux ( like glibc ) which does not provide NODEV, so relying on platform to provide this definition may not be true always.

**Pull request merged**